cudax/include/cuda/experimental/__async/stop_token.cuh

```diff
@@ -0,0 +1,488 @@
+#  include <stop_token>
+#endif
+
+#include <cuda/experimental/__async/prologue.cuh>
+
+// warning #20012-D: __device__ annotation is ignored on a
+// function("inplace_stop_source") that is explicitly defaulted on its first
+// declaration
+_CCCL_NV_DIAG_SUPPRESS(20012)
+
+namespace cuda::experimental::__async
+{
+// [stoptoken.inplace], class inplace_stop_token
+class inplace_stop_token;
+
+// [stopsource.inplace], class inplace_stop_source
+class inplace_stop_source;
+
+// [stopcallback.inplace], class template inplace_stop_callback
+template <class _Callback>
+class inplace_stop_callback;
+
+namespace __stok
+{
+struct __inplace_stop_callback_base
+{
+  _CCCL_HOST_DEVICE void __execute() noexcept
+  {
+    this->__execute_fn_(this);
+  }
+
+protected:
+  using __execute_fn_t = void(__inplace_stop_callback_base*) noexcept;
+
+  _CCCL_HOST_DEVICE explicit __inplace_stop_callback_base( //
+    const inplace_stop_source* __source, //
+    __execute_fn_t* __execute) noexcept
+      : __source_(__source)
+      , __execute_fn_(__execute)
+  {}
+
+  _CCCL_HOST_DEVICE void __register_callback() noexcept;
+
+  friend inplace_stop_source;
+
+  const inplace_stop_source* __source_;
+  __execute_fn_t* __execute_fn_;
+  __inplace_stop_callback_base* __next_      = nullptr;
+  __inplace_stop_callback_base** __prev_ptr_ = nullptr;
+  bool* __removed_during_callback_           = nullptr;
+  _CUDA_VSTD::atomic<bool> __callback_completed_{false};
+};
+
+struct __spin_wait
+{
+  __spin_wait() noexcept = default;
+
+  _CCCL_HOST_DEVICE void __wait() noexcept
+  {
+    if (__count_ == 0)
+    {
+      __async::__this_thread_yield();
+    }
+    else
+    {
+      --__count_;
+      _CUDA_VSTD::__libcpp_thread_yield_processor();
+    }
+  }
+
+private:
+  static constexpr uint32_t __yield_threshold = 20;
+  uint32_t __count_                           = __yield_threshold;
+};
+
+template <template <class> class>
+struct __check_type_alias_exists;
+} // namespace __stok
+
+// [stoptoken.never], class never_stop_token
+struct never_stop_token
+{
+private:
+  struct __callback_type
+  {
+    _CCCL_HOST_DEVICE explicit __callback_type(never_stop_token, __ignore) noexcept {}
+  };
+
+public:
+  template <class>
+  using callback_type = __callback_type;
+
+  _CCCL_HOST_DEVICE static constexpr auto stop_requested() noexcept -> bool
+  {
+    return false;
+  }
+
+  _CCCL_HOST_DEVICE static constexpr auto stop_possible() noexcept -> bool
+  {
+    return false;
+  }
+
+  _CCCL_HOST_DEVICE friend constexpr bool operator==(const never_stop_token&, const never_stop_token&) noexcept
+  {
+    return true;
+  }
+
+  _CCCL_HOST_DEVICE friend constexpr bool operator!=(const never_stop_token&, const never_stop_token&) noexcept
+  {
+    return false;
+  }
+};
+
+template <class _Callback>
+class inplace_stop_callback;
+
+// [stopsource.inplace], class inplace_stop_source
+class inplace_stop_source
+{
+public:
+  _CCCL_HOST_DEVICE inplace_stop_source() noexcept = default;
```

Comment:
```suggestion
  inplace_stop_source() noexcept = default;
```

Reply:
it's there (iirc) because clang-cuda wants it there. i don't know why. i also don't know why i need to add annotations to deduction guides. that's just goofy.